File: mind_snag_py/src/mind_snag/utils/psth.py
"""Peri-stimulus time histogram with Gaussian smoothing.

Ports psth.m. Computes firing rate from trial-aligned spike times.
"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray
from scipy.stats import norm


def psth(
    spike_cell: list[NDArray[np.float64]],
    bn: tuple[int, int],
    smoothing: float = 50.0,
    max_rate: float = 50.0,
) -> tuple[NDArray[np.float64], int]:
    """Compute PSTH from a list of per-trial spike times.

    Parameters
    ----------
    spike_cell : list of arrays, each containing spike times in ms
    bn : (start, stop) time window in ms
    smoothing : Gaussian smoothing std in ms
    max_rate : not used in computation, kept for API compat

    Returns
    -------
    rate : smoothed firing rate vector (spikes/s), length = stop - start + 1
    n_tr : number of trials
    """
    n_tr = len(spike_cell)
    if n_tr == 0:
        return np.zeros(bn[1] - bn[0] + 1), 0

    start, stop = bn
    time_bins = np.arange(start, stop + 1, dtype=np.float64)

    # Collect all spikes shifted by start
    all_spikes = []
    for spikes in spike_cell:
        if spikes is not None and len(spikes) > 0:
            arr = np.asarray(spikes).flatten()
            all_spikes.append(arr)
    if all_spikes:
        xx = np.concatenate(all_spikes)
    else:
        xx = np.array([])

    # Histogram
    z, _ = np.histogram(xx, bins=len(time_bins), range=(start, stop))
    z = z.astype(np.float64)

    # Gaussian smoothing kernel
    half_width = int(3 * smoothing)
    kernel_x = np.arange(-half_width, half_width + 1, dtype=np.float64)
    window = norm.pdf(kernel_x, 0, smoothing)

    # Convolve and trim
    rate = (1000.0 / n_tr) * np.convolve(z, window)
    rate = rate[half_width: half_width + len(time_bins)]

    return rate, n_tr

File: utils/psth.m
function [rate, nTr] = psth(spikecell, bn, smoothing, maxrate, marks, flag)
% PSTH Peri-stimulus time histogram from spike cell array
%
%   [rate, nTr] = psth(spikecell, bn, smoothing, maxrate, marks, flag)
%
%   Parameters:
%     spikecell - Cell array of spike times in ms (starting from 0)
%     bn        - [start, stop] time window in ms
%     smoothing - Gaussian smoothing std in ms (default: 50)
%     maxrate   - Max firing rate for display (default: 50)
%     marks     - Optional marker times for display
%     flag      - If 1, force plotting (default: 0)
%
%   Returns:
%     rate - Smoothed firing rate vector (spikes/s)
%     nTr  - Number of trials

if nargin < 3 || isempty(smoothing); smoothing = 50; end
if nargin < 4; maxrate = 50; end
if nargin < 5; marks = []; end
if nargin < 6; flag = 0; end

nTr = length(spikecell);
dT = maxrate ./ nTr;

Start = bn(1); Stop = bn(2);

XX = [];
Y = [];
for iTr = 1:nTr
    x = spikecell{iTr};
    if size(x,1) == 1; x = x'; end
    x = (x + Start)';
    y = ones(1, length(x)) * (iTr * dT);
    XX = [XX x];
    Y = [Y y];
end

if nargout == 0 || flag == 1
    if nTr > 5
        plot(XX, Y, 'r.', 'MarkerSize', 4);
    else
        plot(XX, Y, 'r.', 'MarkerSize', 10);
    end
end

if ~isempty(marks)
    hold on;
    plot(marks, [1:nTr] .* dT, 'k.', 'Markersize', 10)
end

Z = hist(XX, Start:Stop);
window = normpdf([-3*smoothing:3*smoothing], 0, smoothing);

rate = (1000 / nTr) * conv(Z, window);
rate = rate(3*smoothing + 1 : end - 3*smoothing);

if nargout == 0 || flag == 1
    hold on;
    plot(Start:Stop, rate, 'k', 'linewidth', 3);
    plot([0 0], [0 maxrate], 'b')
    axis([Start Stop 0 maxrate])
end

end
